packages/yarnpkg-pnp/sources/loader/makeApi.ts

```diff
@@ -203,9 +192,51 @@ export function makeApi(runtimeState: RuntimeState, opts: MakeApiOptions): PnpAp
   }
 
   /**
-   * Implements the node resolution for folder access and extension selection
+   * Implements the node resolution for the "exports" field
+   *
+   * @returns The remapped path or `null` if the package doesn't have an "exports" field
    */
+  function applyNodeExportsResolution(unqualifiedPath: PortablePath, {conditions = [], require = true}: ResolveUnqualifiedExportOptions = {}) {
+    const locator = findPackageLocator(ppath.join(unqualifiedPath, `internal.js` as Filename));
+    if (locator === null) {
+      throw makeError(
+        ErrorCode.API_ERROR,
+        `The resolveUnqualifiedExport function must be called with a valid unqualifiedPath`,
+      );
+    }
+
+    const {packageLocation} = getPackageInformationSafe(locator);
+
+    const pkgJson = JSON.parse(opts.fakeFs.readFileSync(ppath.join(packageLocation, `package.json` as Filename), `utf8`));
+
+    let subpath = ppath.contains(packageLocation, unqualifiedPath);
+    if (subpath === null) {
+      throw makeError(
+        ErrorCode.INTERNAL,
+        `unqualifiedPath doesn't contain the packageLocation (this is probably an internal error)`,
+      );
+    }
+
+    if (!/^\.{0,2}\//.test(subpath))
+      subpath = `./${subpath}` as PortablePath;
+
+    const resolvedExport = resolveExport(pkgJson, ppath.normalize(subpath), {
+      browser: false,
+      require,
+      // TODO: implement support for the --conditions flag
+      // Waiting on https://github.com/nodejs/node/issues/36935
+      conditions,
+    });
```

Comment:
I mentioned it on Discord but just to have it here as well so it isn't forgotten:

No matter what we pass here `resolve.exports` [forces two conditions from `['require', 'import', 'browser', 'node']` to always be added](https://github.com/lukeed/resolve.exports/blob/c6814c43cb27ee6092f649be3fdf6c1fb55c51e4/src/index.js#L70-L72) while we need it to only use the conditions we pass so we have full control. For example if `react-native` were to use this to resolve native stuff it should not get matches for neither `node` nor `browser`. While in the ESM loader the conditions can be anything some other loader passes so the same applies there.

cc @lukeed

Reply:
That is correct. See [`defaultConditions`](https://nodejs.org/api/esm.html#esm_resolver_algorithm) and also [here](https://nodejs.org/api/packages.html#packages_conditional_exports). Node is always juggling up to 4 conditions (with 2 forced as defaults). 

You can add whatever custom conditions you want to support, but at the end of the day, the specified key order determines the matching priority. A RN user will need to do something like this for their package to work anywhere (not just _my_ module):

```js
{
  "exports": {
    "react-native": {
      "import": "./rn/index.mjs",
      "require": "./rn/index.cjs"
    },
    "import": "./...",
     //...
  }
}
```

And then specify the "react-native" condition

Reply:
Seems unfortunate that node's defaults should be forced on everybody using `exports` field. For a module that isn't supported to run in Node, it should just fail to resolve or something (or falling back to `default`) not pick up `node`.

There's already a switch in the library for `browser` condition so it's not consistently enforced - why not just fully let the caller provide their own conditions (barring possibly `default` - that seems reasonable to always add)? I.e. remove the two `allows.add` calls (and thus the options leaving only `conditions`). Or probably add `node` if no `conditions` is passed, but otherwise add nothing.

---

I was gonna open up an issue over in `resolve.exports`, but since I came over this PR with this comment before I got to it, I sneakily commented here instead 😀 Happy to open up an issue over there instead

Reply:
As another datapoint, TypeScript could not have used `resolve.exports` to implement support for `types` condition with their official example: https://devblogs.microsoft.com/typescript/announcing-typescript-4-5-beta/#packagejson-exports-imports-and-self-referencing

I think `resolve.exports` should be implementation independent (maybe a `resolve.exports-node` or just a `import { node } from 'resolve.exports'` makes sense, but the base export shouldn't put these limitations there)